phonenumbermatcher.go

```diff
@@ -0,0 +1,388 @@
+	// well.
+	INNER_MATCHES = []*regexp.Regexp{
+		// Breaks on the slash - e.g. "651-234-2345/332-445-1234"
+		regexp.MustCompile("/+(.*)"),
+		// Note that the bracket here is inside the capturing group, since we consider it part of the
+		// phone number. Will match a pattern like "(650) 223 3345 (754) 223 3321".
+		regexp.MustCompile("(\\([^(]*)"),
+		// Breaks on a hyphen - e.g. "12345 - 332-445-1234 is my number."
+		// We require a space on either side of the hyphen for it to be considered a separator.
+		regexp.MustCompile("(?:\\p{Z}-|-\\p{Z})\\p{Z}*(.+)"),
+		// Various types of wide hyphens. Note we have decided not to enforce a space here, since it's
+		// possible that it's supposed to be used to break two numbers without spaces, and we haven't
+		// seen many instances of it used within a number.
+		regexp.MustCompile("[\u2012-\u2015\uFF0D]\\p{Z}*(.+)"),
+		// Breaks on a full stop - e.g. "12345. 332-445-1234 is my number."
+		regexp.MustCompile("\\.+\\p{Z}*([^.]+)"),
+		// Breaks on space - e.g. "3324451234 8002341234"
+		regexp.MustCompile("\\p{Z}+(\\P{Z}+)"),
+	}
+
+	//  Punctuation that may be at the start of a phone number - brackets and plus signs.
+	LEAD_CLASS   = OPENING_PARENS + PLUS_CHARS
+	LEAD_PATTERN = regexp.MustCompile(LEAD_CLASS)
+
+	// Builds the MATCHING_BRACKETS and PATTERN regular expressions. The building blocks below exist to make the pattern more easily understood.
+	OPENING_PARENS = "\\(\\[\uFF08\uFF3B"
+	CLOSING_PARENS = "\\)\\]\uFF09\uFF3D"
+	NON_PARENS     = "[^" + OPENING_PARENS + CLOSING_PARENS + "]"
+
+	// Limit on the number of pairs of brackets in a phone number.
+	BRACKET_PAIR_LIMIT = "{0,3}"
+
+	// Limit on the number of leading (plus) characters.
+	LEAD_LIMIT = "{0,2}"
+
+	// Limit on the number of consecutive punctuation characters.
+	PUNCTIATION_LIMIT = "{0,4}"
+
+	// The maximum number of digits allowed in a digit-separated block. As we allow all digits in a
+	//single block, set high enough to accommodate the entire national number and the international
+	//country code.
+	DIGIT_BLOCK_LIMIT = 17 + 3
+
+	// Limit on the number of blocks separated by punctuation. Uses digitBlockLimit since some
+	// formats use spaces to separate each digit.
+	BLOCK_LIMIT = "{0," + strconv.Itoa(DIGIT_BLOCK_LIMIT) + "}"
+
+	// A punctuation sequence allowing white space.
+	PUNCTUATION = "[" + VALID_PUNCTUATION + "]" + PUNCTIATION_LIMIT
+
+	// A digits block without punctuation.
+	DIGIT_SEQUENCE = "\\d{1," + strconv.Itoa(DIGIT_BLOCK_LIMIT) + "}"
+)
+
+// Creates a new instance.
+//
+// Arguments:
+// text -- The character sequence that we will search
+// country -- The country to assume for phone numbers not written in
+//            international format (with a leading plus, or with the
+//            international dialing prefix of the specified region). May be
+//            "ZZ" if only numbers with a leading plus should be considered.
+func NewPhoneNumberMatcher(text string, region string) PhoneNumberMatcher {
```

Comment:
Looks like it should return the same type as before. And params are different too.

I understand that it wasn't implemented before, but still, everyone should see that this is a breaking change.

CC: @nicpottier

Reply:
I get your point. But I'm not sure what to do here. Currently, the entire PhoneNumberMatcher isn't implemented at all, and the constructor function simply returns nil. 

The new function doesn't do anything exciting, except keeping the struct fields private. If I export the "text", "preferredRegion", "leniency" and "maxTries", PhoneNumberMatcher could just be initialized without constructor function. Is that a better solution?